File: problems/460/solution.go
package main
type Node struct {
    key, val, freq int
    prev, next *Node
}

func NewNode(key, val int) *Node {
    return &Node{
        key, val, 1, nil, nil,
    }
}

type DoublyLinkedList struct {
    head, tail *Node
}

func NewDoublyLinkedList() *DoublyLinkedList {
    head, tail := NewNode(-1, -1), NewNode(-1, -1)
    
    head.next = tail
    tail.prev = head
    
    return &DoublyLinkedList{
        head, tail,
    }
}

func (this *DoublyLinkedList) empty() bool {
    return this.tail.prev == this.head
}

func (this *DoublyLinkedList) add(node *Node) {
    this.head.next.prev = node
    node.next = this.head.next
    node.prev = this.head
    this.head.next = node
}

func (this *DoublyLinkedList) remove(node *Node) {
    prev, next := node.prev, node.next
    
    prev.next = next
    next.prev = prev
}

func (this *DoublyLinkedList) pop() *Node {
    if this.empty() {
        return nil
    }
    
    node := this.tail.prev
    
    this.remove(node)
    
    return node
}

type LFUCache struct {
    cap, size, minFreq int
    nodes map[int]*Node
    freq map[int]*DoublyLinkedList
}

func Constructor(capacity int) LFUCache {
    return LFUCache{
        capacity, 0, 0, make(map[int]*Node), make(map[int]*DoublyLinkedList),
    }
}

func (this *LFUCache) update(node *Node) {
    freq := node.freq
    
    this.freq[freq].remove(node)
    
    if freq == this.minFreq && this.freq[freq].empty() {
        this.minFreq++
    }
    
    node.freq++
    
    if _, ok := this.freq[freq + 1]; !ok {
        this.freq[freq + 1] = NewDoublyLinkedList()
    }
    
    this.freq[freq + 1].add(node)
}

func (this *LFUCache) Get(key int) int {
    if node, ok := this.nodes[key]; ok {
        this.update(node)
        
        return node.val
    }
    
    return -1
}


func (this *LFUCache) Put(key int, value int)  {
    if this.cap == 0 {
        return
    }
    
    if node, ok := this.nodes[key]; ok {
        node.val = value
        this.update(node)
        
        return
    }
    
    if this.size == this.cap {
        toRemove := this.freq[this.minFreq].pop()
        delete(this.nodes, toRemove.key)
        this.size--
    }
    
    node := NewNode(key, value)
    this.nodes[key] = node
    
    if _, ok := this.freq[1]; !ok {
        this.freq[1] = NewDoublyLinkedList()
    }
    
    this.freq[1].add(node)
    this.size++
    this.minFreq = 1
}


/**
 * Your LFUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */

File: problems/460/solution.cpp
class Node {
public:
    int key, val, freq = 1;
    Node *prev = nullptr, *next = nullptr;
    
    Node(int k = -1, int v = -1): key(k), val(v) {}
    Node(int k, int v, Node *p, Node *n): key(k), val(v), next(n), prev(p) {}
};

class DoublyLinkedList {
private:
    Node *head = nullptr, *tail = nullptr;
    
public:
    DoublyLinkedList() {
        head = new Node();
        tail = new Node();
        
        head->next = tail;
        tail->prev = head;
    }
    
    void remove(Node *node) {
        Node *prev = node->prev, *next = node->next;
        next->prev = prev;
        prev->next = next;
    }
    
    void add(Node *node) {
        head->next->prev = node;
        node->next = head->next;
        node->prev = head;
        head->next = node;
    }
    
    Node* pop() {
        if (empty()) {
            return nullptr;
        }
        
        Node *node = tail->prev;
        
        remove(node);
        
        return node;
    }
    
    bool empty() {
        return tail->prev == head;
    }
};

class LFUCache {
private:
    int cap, minFreq = 0, size = 0;
    Node *head = nullptr, *tail = nullptr;
    unordered_map<int, Node*> nodes;
    unordered_map<int, DoublyLinkedList> freq;
    
    void update(Node *node) {
        int nodeFreq = node->freq;
        
        freq[nodeFreq].remove(node);
        
        if (minFreq == nodeFreq and freq[nodeFreq].empty()) {
            minFreq++;
        }
        
        node->freq++;
        freq[nodeFreq + 1].add(node);
    }
    
public:
    LFUCache(int capacity): cap(capacity) {
        head = new Node();
        tail = new Node();
        
        head->next = tail;
        tail->prev = head;
    }
    
    int get(int key) {
        if (not nodes.count(key)) {
            return -1;
        }
        
        update(nodes[key]);
        
        return nodes[key]->val;
    }
    
    void put(int key, int value) {
        if (not cap) {
            return;
        }
        
        if (not nodes.count(key)) {
            if (size == cap) {
                Node *toRemove = freq[minFreq].pop();
                nodes.erase(toRemove->key);
                size--;
            }
            
            Node *node = new Node(key, value);

            nodes[key] = node;
            freq[1].add(node);
            size++;
            minFreq = 1;
            
            return;
        }
        
        update(nodes[key]);
        nodes[key]->val = value;
    }
};

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache* obj = new LFUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */